SEED-platform/seed
# Fix Meter Overwrites

#### Any background context you want to provide?
Inserting a single meter reading was inadvertently deleting any meter readings (for any org) with the same start date.

#### What's this PR do?
- Fixes a major bug where inserting a single meter reading would delete any existing readings for any meter with the same start time
  - Adds tests for this condition
- Adds the ability for inserting bulk meter readings to update existing data
- Sanitizes sql inputs in `copy_readings`

#### How should this be manually tested?
1. Create two meters:

| POST `/api/v3/properties/{property_id}/meters/` |
|---|
| <pre>{<br>  "type": "Electric",<br>  "source": "Manual Entry",<br>  "source_id": "Meter 1"<br>}</pre> |

| POST `/api/v3/properties/{property_id}/meters/` |
|---|
| <pre>{<br>  "type": "Electric",<br>  "source": "Manual Entry",<br>  "source_id": "Meter 2"<br>}</pre> |

2. Add a reading to the first meter:

| POST `/api/v3/properties/{property_id}/meters/{meter_1_id}/readings/` |
|---|
| <pre>{<br>  "start_time": "2023-01-01T00:00:00",<br>  "end_time": "2023-02-01T00:00:00",<br>  "reading": 3.41,<br>  "source_unit": "kWh (thousand Watt-hours)",<br>  "conversion_factor": 3.41<br>}</pre> |

3. Add a reading to the second meter with the same start time:

| POST `/api/v3/properties/{property_id}/meters/{meter_2_id}/readings/` |
|---|
| <pre>{<br>  "start_time": "2023-01-01T00:00:00",<br>  "end_time": "2023-01-02T00:00:00",<br>  "reading": 4559.17,<br>  "source_unit": "kWh (thousand Watt-hours)",<br>  "conversion_factor": 3.41<br>}</pre> |

4. Check that meter 1 still has a value of 1 (not blank)
